### ISSUE TITLE
Calling the library from a Visual Studio C# application

### ISSUE THREAD

**Issue by mr-j0nes** [open-license-manager/open-license-manager#68]
Hi there,

thanks a lot for the great work. I am already using the library on my backend successfully. I need to use the same (or similar library) from my frontend which is in WPF/C#. 

This question goes out to anybody who might have done this already. Any steps to follow on how to do this ? I think it can be wrapped into a managed C++ library, but still have to find out the specifics, if anybody has done this already, it would be much appreciated to get some guidance.

Thanks in advance..

**Comment by gcontini** [comment]
Interesting point. I'm not expert in C#, but bindings to other language are planned (python and java) so this feature request is in the expected development direction. If you can complete the code i'll be very happy to integrate it.

https://docs.microsoft.com/en-us/cpp/dotnet/using-cpp-interop-implicit-pinvoke?view=vs-2019

Most of these interpreted languages require to compile the code as a .dll but since the api is public (the disadvantage of being open source) an attacker could replace the dll with one just returning LICENSE_OK in less than 5 minutes.

**Comment by mr-j0nes** [comment]
Indeed after having had a look into this I see that the only way of making the library available to a C# application is by creating a dll which defeats the whole purpose of having a license file, since anybody with a bit of knowledge could "fake" the functionality and deliver a LICENSE_OK as you say.

I have looked into maybe checking the CRC of the dll being imported, but doesn't look like a good idea.

So for now I guess the only way is porting the library functionalities to C#..

I'd still appreciate if anybody comes up with a different idea..

**Comment by gcontini** [comment]
I would not go as far as closing the issue. It still has value. We have the same issue with Java (and this is an integration we planned to do more sooner than later). 
Of course having interpreted languages makes more easy to bypass the licensing system, however the bypass may be made more difficult eg. obfuscating the names or changing the contract (eg.just changing the value of the enums may help confuse an attacker).
OpenLicenseManager is BSD, nobody prevents you from changing the library in your private git repository to prevent a straight dll substitution.
Also (at least for my experience in Java) java programmers are much less willing to understand a licensing system than C++ developers.

Just the requirement of compiling as dll + the open source constraint makes the release of a generic C# integration layer much harder (not impossible though).

**Comment by emgstabilis** [comment]
One pragmatic solution (not generic, I agree), is to have a core part of
your functionality integrated inside the dll so that just replacing the dll
with a one liner LICENSE_OK is not possible.

Best regards,

Evan Monroig

Le mer. 4 déc. 2019 à 07:17, Gabriele Contini <notifications@github.com> a
écrit :

**Comment by gcontini** [comment]
@emgstabilis yes completely agree. but as you said it's not a generic solution so it is applicable only  to the part of your library that you are going to customize. Can't be published here.

As a course of action we could take in the library is:
1) allow open license manager to be compiled as dll.
2) automagically customize the values in the enum so LICENSE_OK doesn't always correspond to 0 (when we generate the private - public key). 
3) add a magic number in the api that changes the library result (to force an attacker to study a more complex protocol) eg. the application need to pass "42" (configurable) in order to get the license verified. the application may decide to pass different numbers and check the license is not verified.

As a side note a large rearrangment is due in next months (or weeks i hope, the code is already compiling, i'm finishing some integration work) that will change how licenses are issued and managed. When this will be done and "digested" let's see to improve the "automatic project customization" phase (when keys are generated).

**Comment by jmorrison** [comment]
> Interesting point. I'm not expert in C#, but bindings to other language are planned (python and java) so this feature request is in the expected development direction. If you can complete the code i'll be very happy to integrate it.
> 
> https://docs.microsoft.com/en-us/cpp/dotnet/using-cpp-interop-implicit-pinvoke?view=vs-2019
> 
> Most of these interpreted languages require to compile the code as a .dll but since the api is public (the disadvantage of being open source) an attacker could replace the dll with one just returning LICENSE_OK in less than 5 minutes.

Re: bindings - sorry if this is Pilot Error, but the call-by-value of the second argument of type LicenseLocation in acquire_license() is problematic for at least one binding language (Common Lisp).  If I understand correctly, either a pointer or reference to the licenseLocation would let me avoid writing another C layer to bind to.  I suspect this will also be an issue for other language bindings, including Python.  Please advise, and thanks for the software.

**Comment by gcontini** [comment]
@jmorrison I suppose you're talking about the master branch. Please have a look to the develop branch, that should be ok, it pass the tests and it's a good improvement compared to the version in the master branch. So please try 'develop', if it's ok let me know, I will tag and release

**Comment by jmorrison** [comment]
Hi Gabriele;

Thank you for the advice.  I did get it working (as near as I can tell)
under SBCL.  Your modification which eliminated passing-structs-by-value
really, really helped.  Thank you.

You were kind enough further to solicit feedback, so here are a couple
things that might make it easier to generate bindings for dynamic languages
other than Common Lisp (e.g., python):

   - in some structs (e.g., AuditEvent, CallerInformations, LicenseInfo),
   there are fixed-length arrays whose dimensions are not "#defined," but are
   instead manifest constant integers (e.g., "11").  There is also one case
   of  an expression in the dimension (proprietary_data in LicenseInfo).
    These dimensions are thus unavailable as constants into the binding
   language.  So when the binding language tries to set/get these fields,
   there is risk of manually getting the length wrong (especially if it
   changes later), and Bad Things may happen.
   - The concatenation of strings into fixed-length fields in
   CallerInformations may be problematic from some dynamic languages.  I could
   be wrong, but it seemed there was no obvious way to use the "automatically
   coerce Lisp strings to C-strings, returning a pointer" convenience
   functions (for passing to the license verification function), and I had to
   cram characters one-at-a-time into fixed-length arrays in the structure
   (where again, the size of each was unavailable via a c2ffi-provided
   constant).  I presume this may also be the case for python.  I further
   suspect things may get weird in the case of Unicode (not sure, and I did
   not look).

Again, copious thanks for the both the software and your guidance!

-jm

**Comment by gcontini** [comment]
@mr-j0nes  First point accepted (issue #70). I'll take care of it before next release. 
Unfortunately you have to wait some months to see it (because i have other work in progress). If you need it before feel free to issue a pull request. I'll be glad to merge it.

Second point: any proposal?

**Comment by jmorrison** [comment]
Sorry for the delay in replying, as I have been trying to solve
binding-generation issues.

Pursuant to the fixed-length fields in CallerInformations mentioned above,
I further believe that if the project directory filename length exceeds
that fixed project_name limit of 16 characters, then acquire_license
returns a "product not licensed" event type rather than some more
informative diagnostic.  And I have less freedom than I might like as far
as choosing the directory name, as it is used as a "system" name for
dynamically loading packages - it is a namespace pollution issue, wherein
best practice is to prefix a unique name (e.g., my organization names
"com.symsim..." to the package name, making the directory name resemble
"com.symsim.project-name").  I suspect I will not be alone in that regard.

So to return to the second point you asked about, while I do not pretend to
have understood the internal open-license-manager ramifications of
CallerInformations having project_name (and version) as fixed-length
values, it might be useful to either relax the tight (16 char) length
restrictions (use MAX_PATH or the like?), or pass the values indirectly as
pointers/references.  Please advise, and thanks again for all your hard
work!  Please let me know how I can further help the effort.

**Comment by gcontini** [comment]
project_name should not contain any sub folders. 
if you want to configure your project in a specific location use `lcc project initalize` command with -o option then reference the project location in the cmake build using LCC_PROJECTS_BASE_DIR (same dir as -o option) and LCC_PROJECT_NAME

We prefer having fixed lenghts array in case somebody forgets the \0 at the end of the string... The api is much safer.
